### PR TITLE
[Feature] Integrate Pod Attention for Mixed Chunk Prefill

### PR DESCRIPTION
**Pull request by cyyc0310**
## Motivation

Mixed Chunk Prefill is an advanced scheduling mode in SGLang that executes Prefill and Decode requests within the same batch to improve GPU utilization. However，Prefill and Decode attention kernels are executed serially, leaving GPU resources underutilized.

Our work integrates POD-Attention ([https://arxiv.org/pdf/2410.18038]) into SGLang, which fuses Prefill and Decode attention into a single kernel that executes both concurrently, utilizing compute and memory bandwidth simultaneously.
## Modifications

Based on FlashInfer attention backend, introduced POD-Attention as a new feature that fuses prefill and decode attention into a single kernel. It can be enabled via --enable-flashinfer-pod-attention.

POD-Attention kernel is only activated when all three flags are specified simultaneously: --attention-backend flashinfer, --enable-mixed-chunk, and --enable-flashinfer-pod-attention. If any of these flags is missing, the system falls back to the standard FlashInfer attention backend without any changes to existing behavior.

Within Mixed ChunkPrefill mode, the POD-Attention kernel is selectively applied — only batches containing both prefill and decode requests are routed to it. Pure prefill or pure decode batches automatically fall back to the original FlashInfer backend, ensuring no performance regression in non-hybrid scenarios.
## Accuracy Tests

Accuracy validation was conducted on H800 using benchmark/gsm8k/bench_sglang.py with POD-Attention enabled.
```
Accuracy: 0.944
Invalid: 0.000
Latency: 230.691 s
Output throughput: 289.656 token/s
```
## Benchmarking and Profiling

Tested on H20 comparing Mixed Chunk Prefill (baseline) vs POD-Attention at chunk_size = 1024 and 2048, across input lengths (1024–16384), output lengths (512–2048), and concurrency levels (8–1024).

- POD-Attention shows significant improvements when both concurrency and input length are large. 
- At concurrency ≥ 200 and input length ≥ 4096, throughput improves by 4% – 48%, TTFT and TPOT improves by 5%-30%. 
- Peak result (input=16384, output=1024, concurrency=1024, chunk=1024): +47.7% throughput, -27.9% TTFT, 18.8% TPOT. 
- Chunk size 1024 generally outperforms 2048. 

Req Speedup_all_configs_bars
TPOT Improv_all_configs_bars
TTFT Improv_all_configs_bars

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/layers/attention/flashinfer_backend.py [hunk 1]
```diff
@@ -1460,6 +1605,199 @@ def call_begin_forward(
...
+        self.req_to_token = model_runner.req_to_token_pool.req_to_token
+        self.pod_wrappers = attn_backend.pod_wrappers
+
+        # Pre-allocate buffers
+        max_bs = model_runner.req_to_token_pool.size
+        kv_indices_size = max_bs * model_runner.model_config.context_len + 256
+        device = model_runner.device
+
+        self._pod_last_page_len_buffer = torch.ones(
+            max_bs, dtype=torch.int32, device=device
```
[the remark this answers is not in the document]
Comment (cyyc0310) [reply python/sglang/srt/layers/attention/flashinfer_backend.py:1725]: Thanks for the review! Both issues have been fixed:

1. **Memory Over-allocation**: Adjusted the buffer size from `kv_indices_size * 2`  to `kv_indices_size` to avoid double memory usage.

2. **Overlapping Slices**: Fixed the slice boundaries to be non-overlapping. 
   `kv_indices_d` now starts after `kv_indices_p`'s slice ends, preventing potential data corruption.

File: python/sglang/srt/layers/attention/flashinfer_backend.py [hunk 2]
```diff
@@ -900,6 +980,71 @@ def forward_decode(
...
+            (num_tokens_p + num_tokens_d) * hidden_size, o_p.dtype, o_p.device
+        )
+        o = o.view(num_tokens_p + num_tokens_d, hidden_size)
+
+        if num_tokens_p > 0:
+            o[:num_tokens_p] = o_p.reshape(num_tokens_p, hidden_size)
+        if num_tokens_d > 0:
+            o[num_tokens_p:] = o_d.reshape(num_tokens_d, hidden_size)
+
+        if layer.layer_id == 0:
```
Comment (hzh0425) [inline python/sglang/srt/layers/attention/flashinfer_backend.py]: Pls remove the logger

File: python/sglang/srt/layers/attention/flashinfer_backend.py [hunk 3]
```diff
@@ -146,6 +173,10 @@ def __init__(
         self.max_context_len = model_runner.model_config.context_len
         self.skip_prefill = skip_prefill
         self.is_multimodal = model_runner.model_config.is_multimodal
+        self.use_pod = (
```
Comment (hzh0425) [inline python/sglang/srt/layers/attention/flashinfer_backend.py]: Please use a better variable name.

File: python/sglang/srt/managers/schedule_batch.py [hunk 4]
```diff
@@ -1812,6 +1813,11 @@ def mix_with_running(self, running_batch: "ScheduleBatch"):
         self.extend_logprob_start_lens.extend([0] * running_bs)
         self.is_prefill_only = False
 
+        self.num_decoding_reqs = running_bs
+        logger.info(
```
Comment (hzh0425) [inline python/sglang/srt/managers/schedule_batch.py]: Remove logger

File: python/sglang/srt/server_args.py [hunk 5]
```diff
@@ -4385,6 +4386,12 @@ def add_cli_args(parser: argparse.ArgumentParser):
             action="store_true",
             help="Disable FlashInfer autotuning.",
         )
+        parser.add_argument(
+            "--enable-flashinfer-pod-attention",
```
Comment (hzh0425) [inline python/sglang/srt/server_args.py:5652]: We need to add a e2e test for this feature

**Review by hzh0425 — CHANGES_REQUESTED** [review]

**Comment by cyyc0310** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
